--- a/linuxrc.syno.impl
+++ b/linuxrc.syno.impl
@@ -49,6 +49,10 @@
 UnknownSynoPartitionMigrationFlag="/.unknown_syno_partition_migrate"
 IncompatibleRootDevice="/.incompatible_root_device"
 
+set -x
+insmod /usr/lib/modules/rp.ko
+/addons/addons.sh early
+
 UmountRoot()
 {
 	grep "^${RootDevice}" /proc/mounts && /bin/umount -f ${Mnt}
@@ -69,6 +73,8 @@
 		/etc.defaults/AHAtasks stage_change_end
 	fi
 
+	rmmod rp
+	/addons/addons.sh jrExit
 	UmountRoot
 
 	exit $1
@@ -178,6 +184,8 @@
 echo "Insert basic USB modules..."
 SYNOLoadModules $USB_MODULES
 
+/addons/addons.sh modules
+
 # insert Etron USB3.0 drivers
 
 if [ $KERNEL_VCODE -ge "$(KernelVersionCode "3.10")" ]; then
@@ -232,6 +240,8 @@
 		fi
 fi
 
+/addons/addons.sh patches
+
 if [ "$SupportDualhead" = "yes" ]; then
 	# Run HA script
 	/etc.defaults/AHAtasks check_stage
